### PR TITLE
Feature/resonant axis calculator

### PR DESCRIPTION
**Pull request by idovgalyov-4xxi**
Description of the current version of the resonance calculator.

This version works with 2-body resonances for now.
The program contains several modules, main source file and makefile.

The modules are:
global_parameters - containing constants, simple functions and common parameters
id_matrix - module for managing id. matrices (creating, saving in memory, cleaning, etc.)
resonant_axis - calculates resonant values of semimajor axes needed for creating id. matrices (in files)
resonance_finder - finds possible resonances for an asteroid

The main source file "main.f95" is currently using for getting results by "resonance_finder" module.
Makefile is used for compiling and getting executable file "axis".
To run the program use the command `./axis asteroid a delta`, where:
- "asteroid" is an asteroid name
- "a"  - it's semimajor axis (in AU)
- "delta"  - semisize of searching interval (in AU)

The program finds possible resonances and places them in "asteroid.rpin" file.
It also creates files with id. matrices and checks possibility or their use.

Some comments:
- maximum resonance order currently gives as a parameter in "global_parameters" module.
- tests for .fun files are still in progress
- "main.f95" can be changed for testing some procedures if needed.

UPD:
- provided unit tests
- Did some changes in modules corresponding to review notes.

### CODE REVIEW

File: axis/global_parameters.f95 [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    type(idmrow_3body),dimension(:),allocatable:: matrix
+end type idm_3body
+
+type(idm_2body),dimension(9):: idmatrix_2body
+type(idm_3body),dimension(72):: idmatrix_3body
+!----------------------------------------------------------------------------------------------
+contains
+
+!----------------------------------------------------------------------------------------------
+integer function planet_id(s)
```
Comment (smirik) [inline axis/global_parameters.f95]: This function requires a unit test.

File: axis/global_parameters.f95 [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+            i=9
+        case default
+            i=0
+            write(*,*)'Warning! Detected incorrect planet name, be careful.'
+    end select
+    planet_id=i
+end function planet_id
+
+!----------------------------------------------------------------------------------------------
+function planet_name(i) result(s)
```
Comment (smirik) [inline axis/global_parameters.f95]: This function requires a unit test.

File: axis/global_parameters.f95 [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+         case(6)
+             s='SATURN'
+         case(7)
+             s='URANUS'
+         case(8)
+             s='NEPTUNE'
+         case(9)
+             s='PLUTO'
+         case default
+             write(*,*)'Warning! Detected incorrect planet id, be careful.'
```
Comment (smirik) [inline axis/global_parameters.f95]: Better to say:
"Error! The planet id must be between 1 and 9."

File: axis/global_parameters.f95 [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+            i=6
+        case('URANUS')
+            i=7
+        case('NEPTUNE')
+            i=8
+        case('PLUTO')
+            i=9
+        case default
+            i=0
+            write(*,*)'Warning! Detected incorrect planet name, be careful.'
```
Comment (smirik) [inline axis/global_parameters.f95]: Better to say:
"Error! Planet name must be one from the list: MERCURY, VENUS, EARTHMOO, MARS, JUPITER, SATURN, URANUS, NEPTUNE, PLUTO".

File: axis/global_parameters.f95 [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+             s='NEPTUNE'
+         case(9)
+             s='PLUTO'
+         case default
+             write(*,*)'Warning! Detected incorrect planet id, be careful.'
+     end select
+ end function planet_name
+
+!----------------------------------------------------------------------------------------------
+integer function nod(a,b)
```
Comment (smirik) [inline axis/global_parameters.f95]: This function requires a unit test.
Also, don't you think that the function requires too much variables? 
Here are the examples of more easy implementations:
https://users.soe.ucsc.edu/~pohl/FORTRAN/Chfor/gcd.for

File: axis/id_matrix.f95 [hunk 6]
```diff
@@ -0,0 +1,110 @@
+module id_matrix
+
+use global_parameters
+implicit none
+
+
+contains
+
+!----------------------------------------------------------------------------------------------
+integer function get_idmatrix_2body_status(pl_id)
```
Comment (smirik) [inline axis/id_matrix.f95]: This function requires a unit test.

File: axis/global_parameters.f95 (hunk 5 above)
Comment (smirik) [reply axis/global_parameters.f95]: Also it should be named gcd, not nod.

File: axis/id_matrix.f95 [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+!   pl_id - Planet ID
+! Returns:
+!   0 - idmatrix exists and is in RAM
+!  -1 - idmatrix exists only as a file
+!  >0 - idmatrix does not exist
+    character(8):: pl_name
+    integer:: s
+    integer:: pl_id,un
+
+    if(allocated(idmatrix_2body(pl_id)%matrix)) then
```
Comment (smirik) [inline axis/id_matrix.f95]: Could you please explain what happens here?

File: axis/id_matrix.f95 [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+        s=0
```
Comment (smirik) [inline axis/id_matrix.f95]: Isn't it a good idea to use return here instead of continuing the program execution?

File: axis/id_matrix.f95 [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+    else
+        un=8+pl_id
+        pl_name=planet_name(pl_id)
+        open(unit=un,file='id_matrix_'//trim(pl_name)//'.dat',action='read',iostat=s)
+        if(s==0) then
+            close(un)
+            s=-1
```
Comment (smirik) [inline axis/id_matrix.f95]: See the comment above. Isn't return a good option?

File: axis/id_matrix.f95 [hunk 10]
```diff
@@ -0,0 +1,110 @@
...
+
+!----------------------------------------------------------------------------------------------
+subroutine add_idmatrix_2body(pl_id)
+! Loads one id. matrix in RAM from a file (2-body case)
+! Given:
+!   pl_id - planet ID
+! Produces:
+!   idmatrix_2body(pl_id)%matrix
+    integer:: pl_id,s,l,un
+    type(idmrow_2body),dimension(1000):: matrix
```
Comment (smirik) [inline axis/id_matrix.f95]: Why the dimension is 1000.

File: axis/resonance_finder.f95 [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+use global_parameters
+use id_matrix
+implicit none
+
+
+contains
+
+!----------------------------------------------------------------------------------------------
+subroutine get_all_possible_resonances_2body(asteroid,a,delta)
```
Comment (smirik) [inline axis/resonance_finder.f95]: Why is it subroutine, not a function?

File: axis/resonance_finder.f95 [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+                m1=idmatrix_2body(pl_id)%matrix(i)%resonance(1)
+                m= idmatrix_2body(pl_id)%matrix(i)%resonance(2)
+                p1=idmatrix_2body(pl_id)%matrix(i)%resonance(3)
+                p= idmatrix_2body(pl_id)%matrix(i)%resonance(4)
+                if (abs(res_a-a)<=eps) then
+                    write(as,*) pl_name, m1,m,p1,p,res_a,abs(res_a-a)
+                endif
+            enddo
+        case(-1)
+            write(*,*) 'Not found idmatrix in memory - try to find corresponding file...'
```
Comment (smirik) [inline axis/resonance_finder.f95]: Please replace the text with the following:
"Cannot find the identification matrix in the memory.".

File: axis/resonance_finder.f95 [hunk 13]
```diff
@@ -0,0 +1,86 @@
...
+            un=8+pl_id
+            open(unit=un,file='id_matrix_'//trim(pl_name)//'.dat',action='read',iostat=s)
```
Comment (smirik) [inline axis/resonance_finder.f95]: Why do you do here the loading manually, not via invocation of the function: add_idmatrix_2body.

File: axis/resonant_axis.f95 [hunk 14]
```diff
@@ -0,0 +1,111 @@
+module resonant_axis
+
+use global_parameters
+implicit none
+
+contains
+
+!----------------------------------------------------------------------------------------------
+subroutine make_id_matrix_2body(pl_name,max_order)
```
Comment (smirik) [inline axis/resonant_axis.f95]: make -> build
This subroutine should have a unit test (that the file has been created & it is correct).

Comment (smirik) [reply axis/resonant_axis.f95]: Why this subroutine is in resonant_axis.f95, not in id_matrix?

File: axis/resonant_axis.f95 [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+    character(8):: pl_name
+    character(14):: co='(4i4,f23.16)'
+    integer,dimension(4):: resonance
+    
+    pl_id=planet_id(pl_name)
+    un=8+pl_id
+    open(unit=un,file="id_matrix_"//trim(pl_name)//".dat",status='replace')
+    ! Get id. matrix for a planet by a given maximum order
+    do m1=1,max_order
+        do m=-1,-m1-max_order,-1
```
Comment (smirik) [inline axis/resonant_axis.f95]: Why m is negative only? 
5J+2 — IMO, normal resonance, is not it? In your case we check only 5J-2, right?

File: axis/resonant_axis.fun [hunk 16]
```diff
@@ -36,6 +36,17 @@ test n_from_a_reverse_test
...
 
+test count_axis_2body_test
+    real(8):: a1,m1,a
+    integer,dimension(4)::resonance
+    
+    resonance=(/ 5, -2, 0, -3 /) 
+    a1=5.20248019d0
+    m1=9.547919384243222d-4
+    a=count_axis_2body(resonance,a1,m1)
+    assert_equal_within(,2.8253d0,1d-3)
```
Comment (smirik) [inline axis/resonant_axis.fun]: Add, also, please, well-known case for Kirkwood gaps.

**Review by smirik — CHANGES_REQUESTED** [review]
Added comments. Most of them are related to the unit tests, but there are also few comments related to the functions / sub-routines.

File: axis/resonance_finder.f95 (hunk 11 above)
Comment (smirik) [reply axis/resonance_finder.f95]: OK

File: axis/global_parameters.f95 (hunk 4 above)
Comment (idovgalyov-4xxi) [reply axis/global_parameters.f95]: Done

File: axis/global_parameters.f95 (hunk 3 above)
Comment (idovgalyov-4xxi) [reply axis/global_parameters.f95]: Done

File: axis/global_parameters.f95 (hunk 5 above)
Comment (idovgalyov-4xxi) [reply axis/global_parameters.f95]: Refreshed the function, named gcd, deleted 1 variable and modified the algorithm.

File: axis/id_matrix.f95 (hunk 8 above)
Comment (idovgalyov-4xxi) [reply axis/id_matrix.f95]: Modified this function for using "return" with no conflicts with Fortran specifications.

File: axis/id_matrix.f95 (hunk 10 above)
Comment (idovgalyov-4xxi) [reply axis/id_matrix.f95]: Fixed this.
The method I've described.
Thought it would be acceptable.

File: axis/resonance_finder.f95 (hunk 12 above)
Comment (idovgalyov-4xxi) [reply axis/resonance_finder.f95]: Done

File: axis/resonance_finder.f95 (hunk 13 above)
Comment (idovgalyov-4xxi) [reply axis/resonance_finder.f95]: Justifiable.
Applied this subroutine.

File: axis/resonant_axis.f95 (hunk 14 above)
Comment (idovgalyov-4xxi) [reply axis/resonant_axis.f95]: Renamed and moved to "id_matrix" module.
There are still questions about unit test...

File: axis/resonant_axis.f95 (hunk 15 above)
Comment (idovgalyov-4xxi) [reply axis/resonant_axis.f95]: Found some precedents.
Left as is for now.

File: axis/resonant_axis.fun (hunk 16 above)
Comment (idovgalyov-4xxi) [reply axis/resonant_axis.fun]: Done

File: axis/id_matrix.f95 (hunk 6 above)
Comment (idovgalyov-4xxi) [reply axis/id_matrix.f95]: There are still questions about unit test...

**Review by smirik — APPROVED** [review]

**Review by smirik — CHANGES_REQUESTED** [review]
The tests are still not working, even if replace .f95 with .f90. The issue is in REAL(4) vs REAL(8). Need to be fixed before the commit.

File: axis/global_parameters.fun [hunk 17]
```diff
@@ -0,0 +1,54 @@
...
+    pl=planet_anrt_name(p1)
+    p=planet_id(pl)
+    assert_equal(p1,p)
+end test
+
+test gcd_test
+    integer :: a1=5,b1=2,g1=1
+    integer :: a2=60,b2=12,g2=12
+    integer :: g
+    g=gcd(a1,b1)
```
Comment (smirik) [inline axis/global_parameters.fun]: IMO, additional things.
Why not to check directly?
`assert_equal(12,gcd(60,12))`?
Why do you create an extra variables?

File: axis/id_matrix.f95 [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+            l = l + 1
+        enddo
+        rewind (un)
+
+        allocate (idmatrix_2body(pl_id)%matrix(1:l))
+        do i = 1, l
+            read (un, *) idmatrix_2body(pl_id)%matrix(i)
+        enddo
+        close (un)
+! TO DO - here is a weakness of the program with uncertainity of matrix length - need a cure
```
Comment (smirik) [inline axis/id_matrix.f95]: Why the code is commented?

File: axis/id_matrix.fun [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    
+    call build_idmatrix_2body(pl_name,max_order)
+    open(unit=108,file='id_matrix_JUPITER.dat',action='read',iostat=s)
+    assert_true(s==0)
+end test
+
+
+end test_suite
+
+! assert_true(expression)
```
Comment (smirik) [inline axis/id_matrix.fun]: These strings with the comments were added initially as examples. Please remove them!

File: axis/global_parameters.fun [hunk 20]
```diff
@@ -0,0 +1,54 @@
...
+    assert_equal(g1,g)
+    g=gcd(a2,b2)
+    assert_equal(g2,g)
+end test
+
+end test_suite
+
+! assert_true(expression)
```
Comment (smirik) [inline axis/global_parameters.fun]: These strings with the comments were added initially as examples. Please remove them!

File: axis/id_matrix.f95 (hunk 18 above)
Comment (smirik) [inline axis/id_matrix.f95]: Why the code is commented?

File: axis/global_parameters.fun (hunk 17 above)
Comment (smirik) [inline axis/global_parameters.fun]: IMO, additional things.
Why not to check directly?
`assert_equal(12,gcd(60,12))`?
Why do you create an extra variables?

File: axis/main.f95 [hunk 21]
```diff
@@ -0,0 +1,43 @@
+program resonant_axis_calculator
```
Comment (smirik) [inline axis/main.f95]: Just FYI: if I run the command without the arguments, it shows the error. But how can I find the arguments I should put?

File: axis/resonance_finder.f95 [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+                m1 = idmatrix_2body(pl_id)%matrix(i)%resonance(1)
+                m = idmatrix_2body(pl_id)%matrix(i)%resonance(2)
+                p1 = idmatrix_2body(pl_id)%matrix(i)%resonance(3)
+                p = idmatrix_2body(pl_id)%matrix(i)%resonance(4)
+                if (abs(res_a - a) <= eps) then
+                    write (as, *) pl_name, m1, m, p1, p, res_a, abs(res_a - a), eps
+                endif
+            enddo
+        case default
+            write (*, *) 'No founded idmatrix for ', pl_name, '!'
```
Comment (smirik) [inline axis/resonance_finder.f95]: 'No founded idmatrix' -> 'idmatrix was not found'

**Review by smirik — CHANGES_REQUESTED** [review]
See the comments

File: axis/id_matrix.fun (hunk 19 above)
Comment (idovgalyov-4xxi) [reply axis/id_matrix.fun]: Done

File: axis/id_matrix.f95 (hunk 18 above)
Comment (idovgalyov-4xxi) [reply axis/id_matrix.f95]: Here I decided to use more stable file-measuring method.
I came to decision that it wouldn't be a significant fall of performance.

File: axis/main.f95 (hunk 21 above)
Comment (idovgalyov-4xxi) [reply axis/main.f95]: Provided Readme with description of how to lsunch the current version.

File: axis/resonance_finder.f95 (hunk 22 above)
Comment (idovgalyov-4xxi) [reply axis/resonance_finder.f95]: Done

**Review by smirik — APPROVED** [review]

**Pull request merged**